discojs/src/models/gpt/layers.spec.ts

```diff
@@ -0,0 +1,308 @@
+    //     const outData = await output.data();
+    //     // Expected calculation:
+    //     // [1,2] dot [[1,0],[0,1]] = [1,2] and then add bias [0.5, -0.5] gives [1.5, 1.5]
+    //     expect(Array.from(outData)).to.deep.equal([1.5, 1.5]);
+    //   });
+    // });
+  
+    describe('splitHeads', function () {
+      it('should reshape and transpose the input correctly', function () {
+        const B = 2;
+        const T = 6;
+        const totalChannels = config.nEmbd; // 8 channels
+        // input tensor with shape [B, T, totalChannels]
+        const input = tf.tensor3d(new Array(B * T * totalChannels).fill(1), [B, T, totalChannels]);
+        const output = csa.splitHeads(input, B, T, config.nHead);
+        // expected shape: [B, nHead, T, totalChannels/nHead] = [2, 2, 6, 4]
+        expect(output.shape).to.deep.equal([B, config.nHead, T, totalChannels / config.nHead]);
+      });
+    });
+  
+    describe('applyCausalMask', function () {
+      it('should produce a causal mask that sets upper-triangular positions to -1e9', async function () {
+        const T = config.contextLength;
+        // dummy attention logits tensor with shape [1, 1, T, T] filled with zeros
+        const att = tf.zeros([1, 1, T, T], 'float32');
+        const masked = csa.applyCausalMask(att, T);
+        const data = await masked.data();
+        // for each position (i,j): if j > i expect -1e9 else 0
+        const expected: number[] = [];
+        for (let i = 0; i < T; i++) {
+          for (let j = 0; j < T; j++) {
+            expected.push(j > i ? -1e9 : 0);
+          }
+        }
+        expect(Array.from(data)).to.deep.equal(expected);
+      });
+    });
+  
+    describe('computeAttention', function () {
+      it('should output attention weights that sum to 1 over the last dimension', async function () {
+        const B = 1;
+        const nHead = config.nHead;
+        const T = config.contextLength;
+        const headSize = config.nEmbd / config.nHead;
+        const q = tf.randomUniform([B, nHead, T, headSize]);
+        const k = tf.randomUniform([B, nHead, T, headSize]);
+        const att = csa.computeAttention(q, k, false, T);
+        // expected shape: [B, nHead, T, T]
+        expect(att.shape).to.deep.equal([B, nHead, T, T]);
+        // check that each row of the attention logits (last dimension) sums to approximately 1
+        const attData = await att.data();
+        const attArray = Array.from(attData);
+        for (let b = 0; b < B; b++) {
+          for (let h = 0; h < nHead; h++) {
+            for (let i = 0; i < T; i++) {
+              // calculate the starting index for the i-th row in the flattened tensor
+              const rowStart = b * nHead * T * T + h * T * T + i * T;
+              const row = attArray.slice(rowStart, rowStart + T);
+              const rowSum = row.reduce((sum, val) => sum + val, 0);
+              expect(rowSum).to.be.closeTo(1, 1e-3);
+            }
+          }
+        }
```

Comment:
you can [sum over the last dimension](https://js.tensorflow.org/api/4.22.0/#sum) ("-1" mean the last one if tfjs)

```suggestion
        for (const rowSum of await att.sum(-1).data())
          expect(rowSum).to.be.closeTo(1, 1e-3);
```